### PR TITLE
feat(elements): Consider ValidityState in FieldState

### PR DESCRIPTION
**Pull request by LekoArts**
## Description

This PR adds the ValidityState (https://www.radix-ui.com/primitives/docs/components/form#validitystate) to our FieldState logic. So e.g. if someone enters an invalid email address into a `<input type="email" />` and focus away, our `<FieldState>` will now report an `error`. Previously it didn't do anything.

In this PR also some data attributes were removed/renamed, see changeset + inline comments for more details.

Docs PR: https://github.com/clerk/clerk-docs/pull/1179

Fixes SDK-1638

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [x] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/elements/src/react/common/form/index.tsx [hunk 1]
```diff
@@ -149,7 +166,7 @@ const useForm = ({ flowActor }: { flowActor?: BaseActorRef<{ type: 'SUBMIT' }> }
 
   return {
     props: {
-      [`data-${validity}`]: true,
+      ...(errors.length > 0 ? { 'data-global-error': true } : {}),
```
Comment (LekoArts) [inline packages/elements/src/react/common/form/index.tsx:169]: It didn't make much sense to me to have `data-valid`/`data-invalid` on the `<form>` if that's already placed further below in the DOM tree. And in those instances we refer to the `<FieldError>` state, not `<GlobalError>`.

So renaming this to make it more clear that this is about global errors.

File: packages/elements/src/react/common/form/index.tsx [hunk 2]
```diff
@@ -161,12 +178,10 @@ const useField = ({ name }: Partial<Pick<FieldDetails, 'name'>>) => {
 
   const shouldBeHidden = false; // TODO: Implement clerk-js utils
   const hasError = feedback ? feedback.type === 'error' : false;
-  const validity = hasError ? FIELD_VALIDITY.invalid : FIELD_VALIDITY.valid;
 
   return {
     hasValue,
     props: {
-      [`data-${validity}`]: true,
```
Comment (LekoArts) [inline packages/elements/src/react/common/form/index.tsx:169]: Radix already adds `data-valid`/`data-invalid` correctly to the field, so this is removed to not end up in a situation where both can exist at the same time.

File: packages/elements/src/react/common/form/index.tsx [hunk 3]
```diff
@@ -443,7 +459,17 @@ const FieldInner = React.forwardRef<FormFieldElement, FormFieldProps>((props, fo
...
+        {validity => {
+          const enrichedFieldState = enrichFieldState(validity, fieldState);
+
+          return (
+            <ValidityStateContext.Provider value={validity}>
+              {typeof children === 'function' ? children(enrichedFieldState) : children}
+            </ValidityStateContext.Provider>
+          );
+        }}
+      </RadixValidityState>
```
Comment (LekoArts) [inline packages/elements/src/react/common/form/index.tsx:472]: Main gist of this PR. Using Context here to not have to pass around values through props 👍

File: packages/elements/src/react/common/form/index.tsx [hunk 4]
```diff
@@ -89,6 +77,36 @@ const determineInputTypeFromName = (name: FormFieldProps['name']) => {
...
 };
 
+/**
+ * Radix can return the ValidityState object, which contains the validity of the field. We need to merge this with our existing fieldState.
+ * When the ValidityState is valid: false, the fieldState should be overriden. Otherwise, it shouldn't change at all.
+ * @see https://www.radix-ui.com/primitives/docs/components/form#validitystate
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/ValidityState
+ */
+const enrichFieldState = (validity: ValidityState | undefined, fieldState: FieldStates) => {
+  return validity?.valid === false ? FIELD_STATES.error : fieldState;
```
[the remark this answers is not in the document]
Comment (LekoArts) [reply packages/elements/src/react/common/form/index.tsx:87]: I also thought about doing this instead:

```ts
return fieldState === FIELD_STATES.idle ? validity?.valid : fieldState
```

I guess it's a question of what should take precedence. The validity or our own fieldState

**Review by alexcarpenter — APPROVED** [review]

**Review by tmilewski — APPROVED** [review]

**Pull request merged**